### PR TITLE
Feature/validator enum default

### PR DESCRIPTION
**Pull request by ralphdoe**
### Description

Changed the way this validation worked, the enum validations are related to schemas, this covers both, parameters and schemas for OAS 2 and 3. 

### How Has This Been Tested?

Old tests moved to schemas.js and added new tests.

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [x] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

File: test/unit/plugins/validate-semantic/2and3/schemas.js [hunk 1]
```diff
@@ -543,4 +543,232 @@ describe(`values in Enum must be instance of the defined type`, () => {
...
+      return validateHelper(spec)
+        .then(system => {
+          const allErrors = system.errSelectors.allErrors().toJS()
+          const firstError = allErrors[0]
+          expect(allErrors.length).toEqual(1)
+          expect(firstError.path).toEqual(["definitions", "MySchema","default"])
+          expect(firstError.message).toEqual("Default values must be present in `enum`")
+        })
+    })
+    it("should return an error for an invalid OpenAPI 3 definition", () => {
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/2and3/schemas.js:611]:
```suggestion
    it("should return an error for an invalid OpenAPI 3 parameter definition", () => {
```

File: test/unit/plugins/validate-semantic/2and3/schemas.js [hunk 2]
```diff
@@ -543,4 +543,232 @@ describe(`values in Enum must be instance of the defined type`, () => {
...
+      return validateHelper(spec)
+        .then(system => {
+          const allErrors = system.errSelectors.allErrors().toJS()
+          const firstError = allErrors[0]
+          expect(allErrors.length).toEqual(1)
+          expect(firstError.path).toEqual(["paths", "/pets", "get", "parameters", "0", "schema", "default"])
+          expect(firstError.message).toEqual("Default values must be present in `enum`")
+        })
+    })
+    it("should return an error for an invalid OpenAPI 3 definition", () => {
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/2and3/schemas.js:642]:
```suggestion
    it("should return an error for an invalid OpenAPI 3 parameter component", () => {
```

File: src/plugins/validate-semantic/validators/2and3/schemas.js [hunk 3]
```diff
@@ -68,3 +68,31 @@ export const validate2And3TypesInDefaultValuesMatchesWithEnum = () => (system) =
...
+      return nodes.reduce((acc, node) => {
+        const element = node.node || {}
+        let elementEnum, elementDefault, internalLocation
+
+        if(!element || element.enum === undefined || element.default === undefined) {
+          // nothing to do
+          return acc
+        }
+        elementEnum = element.enum
+        elementDefault = element.default
```
Comment (hkosova) [inline src/plugins/validate-semantic/validators/2and3/schemas.js:85]: The `elementEnum` and `elementDefault` variables are used only once. You can either

a) remove them and change the 2nd `if` to
```js
        if(element.enum.indexOf(element.default) === -1) {
```

or

b) assign them earlier and change the 1st `if` to
```js
        if(!element || elementEnum === undefined || elementDefault === undefined) {
```

File: src/plugins/validate-semantic/validators/2and3/schemas.js [hunk 4]
```diff
@@ -68,3 +68,31 @@ export const validate2And3TypesInDefaultValuesMatchesWithEnum = () => (system) =
...
+        internalLocation = ["default"]
+
+        if(elementEnum.indexOf(elementDefault) === -1) {
+          acc.push({
+            message: "Default values must be present in `enum`",
+            path: [...node.path, ...internalLocation]
```
[the remark this answers is not in the document]
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/2and3/schemas.js:91]: Done!

File: test/unit/plugins/validate-semantic/2and3/schemas.js [hunk 5]
```diff
@@ -543,4 +543,232 @@ describe(`values in Enum must be instance of the defined type`, () => {
       expect(allErrors.length).toEqual(0)
     })
   })
+
+  describe(`schema defaults must be present in enums`, () => {
+    it("should return an error for an invalid Swagger 2 definition", () => {
```
[the remark this answers is not in the document]
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/2and3/schemas.js:548]: Done!

File: test/unit/plugins/validate-semantic/2and3/schemas.js [hunk 6]
```diff
@@ -543,4 +543,232 @@ describe(`values in Enum must be instance of the defined type`, () => {
...
+      return validateHelper(spec)
+        .then(system => {
+          const allErrors = system.errSelectors.allErrors().toJS()
+          const firstError = allErrors[0]
+          expect(allErrors.length).toEqual(1)
+          expect(firstError.path).toEqual(["paths", "/pets", "get", "parameters", "0", "default"])
+          expect(firstError.message).toEqual("Default values must be present in `enum`")
+        })
+    })
+    it("should return an error for an invalid Swagger 2 definition", () => {
```
[the remark this answers is not in the document]
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/2and3/schemas.js:577]: Done!

File: test/unit/plugins/validate-semantic/2and3/schemas.js (hunk 1 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/2and3/schemas.js:611]: Done!

File: test/unit/plugins/validate-semantic/2and3/schemas.js (hunk 2 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/2and3/schemas.js:642]: Done!

File: src/plugins/validate-semantic/validators/2and3/schemas.js (hunk 3 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/2and3/schemas.js:85]: Done, took the first option.

**Comment by hkosova** [comment]
please build

**Comment by tim-lai** [comment]
@ralphdoe Approved. First, please update with latest master, thanks.
